### ISSUE TITLE
New Subject Identifier standard

### ISSUE THREAD

**Issue by AndersAbel** [Sustainsys/Saml2#1040]
There is a new subject identifier standard http://docs.oasis-open.org/security/saml-subject-id-attr/v1.0/csprd03/saml-subject-id-attr-v1.0-csprd03.pdf that needs to be looked into. It needs some fallback processing to produce a valid subject identifier to hand off for e.g. asp.net identity.

**Comment by robertpmansion** [comment]
@AndersAbel  can be also considered as reference: [https://kantarainitiative.github.io/SAMLprofiles/saml2int.html](https://kantarainitiative.github.io/SAMLprofiles/saml2int.html)

**Comment by AndersAbel** [comment]
@robertpmansion Yes, that's actually how I found out about it. Are you further familiar with it?

**Comment by rpmansion** [comment]
@AndersAbel I had a conversation I believe last year with one of the federation we're subscribing regarding about this. I'm planning to contact them again and cc you to further discuss this standard. How does that sound?

**Comment by AndersAbel** [comment]
@rpmansion Sounds great. Understanding how various mechanisms are used in real world scenarios is a key to implement the right features. We could spend years writing a complete implementation off the specs - and have 80% wasted work.

**Comment by rpmansion** [comment]
@AndersAbel Yes, I will try to contact them this week. We can also add this as enhancement for v3. :)

**Comment by rpmansion** [comment]
From UK Federation:

> The latest version of the spec is at https://docs.oasis-open.org/security/saml-subject-id-attr/v1.0/cs01/saml-subject-id-attr-v1.0-cs01.html. The link you have references an earlier version, and the specification was granted full specification status in 2019. I don't recall any significant technical difference between the csprd02, csprd03 and cs01 versions, but you should use the latest one. Thankfully, it's a stable one.
> 
> ...
> 
> However, there are few production uses for the identifier at the moment. Only SWITCH -- the Swiss federation -- has any production entities using it. However, that's what we'd expect from new specifications: software implementations need to support the feature in tandem with federation operators supporting it.
> 
> I'd still urge you to consider implementing the new identifiers. If the sustainsys software has a modular approach to decoding subject identifiers and passing to the application environment, it shouldn't be a difficult implementation task. What will be more tricky is the task of migrating accounts from old identifiers (such as eduPersonTargetedID and eduPersonPrincipalName) to the new identifiers. 

> ...
> 
> I'm glad to see that the sustainsys code is open source. There appears to be existing code that should be able to extract the new attributes containing subject identifiers, at https://github.com/Sustainsys/Saml2/blob/ad39f77c7c55279231d12d601c115f01a76f8ff4/Sustainsys.Saml2/Saml2StatementExtension.cs#L60 which would already be used for current identifiers like eduPersonTargetedID and eduPersonPrincipalName. I don't see anything in that repository which defines the existing identifiers, so I'm presuming that the handling of attributes is at the application level, and that's probably where any development work to support and transition to the new identifiers would have to happen.

**Comment by rpmansion** [comment]
@AndersAbel, including your response to UK Federation below for reference.

> I think that there are three steps that need to be looked into to get the Sustainsys.Saml2 library compliant with the new standard.
> 1. The metadata structure needed to support this through an EntityAttribute. I think it might be possible to add the required metadata through current hooks/callbacks, but in that case it requires composing the XML node tree manually.
> 2. The new attribute should be treated as the main subject identifier. This can definitely be done as everything is just transformed by the library to a list of claims. The AcsCommandResultCreated callback can modify the generated claims list to use one of the new attributes as the main subject identifier instead of the classical NameId.
> 3. Allow assertions without a NameID. I think that these are currently blocked/rejected by the library.
> 
> For now I think that metadata support for EntityAttrbitues (and EntityCategories which are needed by some federations for other purposes) is the most pressing thing to implement.
